Lib/typing.py

```diff
@@ -1675,13 +1675,65 @@ class _TypingEllipsis:
 
 _SPECIAL_NAMES = frozenset({
     '__abstractmethods__', '__annotations__', '__dict__', '__doc__',
-    '__init__', '__module__', '__new__', '__slots__',
-    '__subclasshook__', '__weakref__', '__class_getitem__',
-    '__match_args__',
+    '__module__', '__slots__', '__match_args__', '__qualname__',
+})
+_SPECIAL_CALLABLE_NAMES = frozenset({
+    '__init__', '__new__', '__subclasshook__','__class_getitem__', '__weakref__',
 })
 
 # These special attributes will be not collected as protocol members.
 EXCLUDED_ATTRIBUTES = _TYPING_INTERNALS | _SPECIAL_NAMES | {'_MutableMapping__marker'}
+EXCLUDED_MEMBERS = EXCLUDED_ATTRIBUTES | _SPECIAL_CALLABLE_NAMES
+
+
+def _get_local_members(namespace):
+    """Collect the specified attributes from a classes' namespace."""
+    annotations = namespace.get("__annotations__", {})
+    attrs = (namespace.keys() | annotations.keys()) - EXCLUDED_MEMBERS
+    # exclude special "_abc_" attributes
+    return {attr for attr in attrs if not attr.startswith('_abc_')}
+
+
+def _get_local_protocol_members(namespace):
+    """Collect the specified attributes from the protocols' namespace."""
+    # annotated attributes are always considered protocol members
+    annotations = namespace.get("__annotations__", {})
+    # only namespace members outside the excluded set are considered protocol members
+    return (namespace.keys() - EXCLUDED_ATTRIBUTES) | annotations.keys()
+
+
+def _get_parent_members(cls):
+    """Collect protocol members from parents of arbitrary class object.
+
+    This includes names actually defined in the class dictionary, as well
+    as names that appear in annotations. Special names (above) are skipped.
+    """
+    attrs = set()
+    for base in cls.__mro__[1:-1]:  # without self and object
+        if base.__name__ in {'Protocol', 'Generic'}:
+            continue
+        elif getattr(base, "_is_protocol", False):
+            attrs |= getattr(base, "__protocol_attrs__", set())
+        else:  # get from annotations
+            attrs |= _get_local_members(base.__dict__)
+    return attrs
+
+
+def _get_cls_members(cls):
+    """Collect protocol members from an arbitrary class object.
+
+    This includes names actually defined in the class dictionary, as well
+    as names that appear in annotations. Special names (above) are skipped.
+    """
+    attrs = set()
+    for base in cls.__mro__[:-1]:  # without object
+        if base.__name__ in {'Protocol', 'Generic'}:
+            continue
+        elif getattr(base, "_is_protocol", False):
+            attrs |= getattr(base, "__protocol_attrs__", set())
+        else:
+            attrs |= _get_local_members(base.__dict__)
+    return attrs
 
 
 def _get_protocol_attrs(cls):
```

Comment:
After the modifications here, this function is not used anymore, and it will even give some wrong results (e.g. if the Protocol defines `__class_getitem__`).

Reply:
If it's not used anywhere anymore, then yes, it should be deleted. It's an undocumented, private function; if anybody's using it and their code is broken by us deleting it, that's on them. I can't find any uses of it on grep.app anyway, so I don't think anybody _will_ have their code broken by us deleting it.